internal/server/server_test.go

```diff
@@ -0,0 +1,245 @@
+	cfg := &config.Config{
+		Server: config.ServerConfig{
+			Address: "localhost",
+			Port:    8080,
+		},
+	}
+
+	registry := toolsets.NewRegistry(cfg, []toolsets.Toolset{})
+
+	srv := NewServer(cfg, registry)
+
+	require.NotNil(t, srv)
+	assert.Equal(t, cfg, srv.cfg)
+	assert.Equal(t, registry, srv.registry)
+	assert.NotNil(t, srv.mcp)
+}
+
+func TestServer_registerTools_AllEnabled(t *testing.T) {
+	cfg := &config.Config{
+		Global: config.GlobalConfig{
+			ReadOnlyTools: false,
+		},
+		Server: config.ServerConfig{
+			Address: "localhost",
+			Port:    8080,
+		},
+	}
+
+	readOnlyTestTool := &mockTool{name: "test_read_only_tool", readOnly: true}
+	readWriteTestTool := &mockTool{name: "test_read_write_tool", readOnly: false}
+
+	toolsetList := []toolsets.Toolset{
+		newMockToolset("test_toolset", true, []toolsets.Tool{readOnlyTestTool, readWriteTestTool}),
+	}
+
+	registry := toolsets.NewRegistry(cfg, toolsetList)
+	srv := NewServer(cfg, registry)
+
+	srv.registerTools()
+
+	assert.True(t, readOnlyTestTool.registerCall, "read-only test tool should be registered")
+	assert.True(t, readWriteTestTool.registerCall, "read-write test tool should be registered")
+}
+
+func TestServer_registerTools_ReadOnlyMode(t *testing.T) {
+	cfg := &config.Config{
+		Global: config.GlobalConfig{
+			ReadOnlyTools: true,
+		},
+		Server: config.ServerConfig{
+			Address: "localhost",
+			Port:    8080,
+		},
+	}
+
+	readOnlyTestTool := &mockTool{name: "test_read_only_tool", readOnly: true}
+	readWriteTestTool := &mockTool{name: "test_read_write_tool", readOnly: false}
+
+	toolsetList := []toolsets.Toolset{
+		newMockToolset("test_toolset", true, []toolsets.Tool{readOnlyTestTool, readWriteTestTool}),
+	}
+
+	registry := toolsets.NewRegistry(cfg, toolsetList)
+	srv := NewServer(cfg, registry)
+
+	srv.registerTools()
+
+	assert.True(t, readOnlyTestTool.registerCall, "read-only test tool should be registered")
+	assert.False(t, readWriteTestTool.registerCall, "read-write test tool should not be registered in read-only mode")
+}
+
+func TestServer_registerTools_DisabledToolset(t *testing.T) {
+	cfg := &config.Config{
+		Global: config.GlobalConfig{
+			ReadOnlyTools: false,
+		},
+		Server: config.ServerConfig{
+			Address: "localhost",
+			Port:    8080,
+		},
+	}
+
+	enabledTestTool := &mockTool{name: "test_enabled_tool", readOnly: true}
+	disabledTestTool := &mockTool{name: "test_disabled_tool", readOnly: true}
+
+	toolsetList := []toolsets.Toolset{
+		newMockToolset("enabled_toolset", true, []toolsets.Tool{enabledTestTool}),
+		newMockToolset("disabled_toolset", false, []toolsets.Tool{disabledTestTool}),
+	}
+
+	registry := toolsets.NewRegistry(cfg, toolsetList)
+	srv := NewServer(cfg, registry)
+
+	srv.registerTools()
+
+	assert.True(t, enabledTestTool.registerCall, "tool from enabled toolset should be registered")
+	assert.False(t, disabledTestTool.registerCall, "tool from disabled toolset should not be registered")
+}
+
+func TestServer_Start(t *testing.T) {
```

Comment:
This is a bit redundant what we are testing in `main_test.go`, so we should consider is there anything better we can do here?